src/monty/io.py

```diff
@@ -14,40 +14,89 @@
 import os
 import subprocess
 import time
+import warnings
 from pathlib import Path
 from typing import TYPE_CHECKING
 
 if TYPE_CHECKING:
-    from typing import IO, Generator, Union
+    from typing import IO, Any, Generator, Union
 
 
-def zopen(filename: Union[str, Path], *args, **kwargs) -> IO:
+def zopen(
+    filename: Union[str, Path],
+    /,
+    mode: str | None = None,
+    **kwargs: Any,
+) -> IO:  # TODO: incomplete return type for compressed files
     """
-    This function wraps around the bz2, gzip, lzma, xz and standard Python's open
-    function to deal intelligently with bzipped, gzipped or standard text
-    files.
+    This function wraps around `[bz2/gzip/lzma].open` and `open`
+    to deal intelligently with compressed or uncompressed files.
+    Supports context manager:
+        `with zopen(filename, mode="rt", ...)`.
+
+    Important Notes:
+        - Default `mode` should not be used, and would not be allow
+            in future versions.
+        - Always explicitly specify binary/text in `mode`, i.e.
+            always pass `t` or `b` in `mode`.
+        - When using text mode, always provide an explicit encoding.
 
     Args:
-        filename (str/Path): filename or pathlib.Path.
-        *args: Standard args for Python open(..). E.g., 'r' for read, 'w' for
-            write.
-        **kwargs: Standard kwargs for Python open(..).
+        filename (str | Path): Filename.
+        mode (str): The mode in which the file is opened (
+            e.g., "r" for read, "w" for write.), you MUST
+            explicitly specify "b" for binary or "t" for text.
+        **kwargs: Additional keyword arguments to pass to `open`.
 
     Returns:
-        File-like object. Supports with context.
+        TextIO | BinaryIO
     """
-    if filename is not None and isinstance(filename, Path):
-        filename = str(filename)
-
-    _name, ext = os.path.splitext(filename)
-    ext = ext.upper()
-    if ext == ".BZ2":
-        return bz2.open(filename, *args, **kwargs)
-    if ext in {".GZ", ".Z"}:
```

Comment:
I don't think Python's built-in `gzip` could open a [**real LZW compressed file**](https://en.wikipedia.org/wiki/Gzip) (not a file compressed by `gzip` DEFLATE algorithm and name to `.z`):
> The gzip file format is also not to be confused with that of the compress utility, based on LZW, with extension .Z

For example:
```python
import gzip
import os
import subprocess


with open("hello.txt", mode="wt", encoding="utf-8") as file:
    file.write("hello\n")

subprocess.run(["compress", "-f", "hello.txt"], check=True)
assert os.path.isfile("hello.txt.Z")

with gzip.open("hello.txt.Z", mode="rt") as file:
    print(file.readlines())  # gzip.BadGzipFile: Not a gzipped file (b'\x1f\x9d')
```